### DISCUSSION TITLE
Indicate scores set in osu!stable vs scores set with the Classic mod in osu!lazer

### DISCUSSION THREAD

**Discussion by Natelytle** [ppy/osu#22739] [Ideas]
Classic scores set in lazer have differences to scores set in stable, such as information on slider ends dropped and sliderticks missed, as well as not losing combo from a dropped sliderend. In my opinion, it would make sense UX wise and for the purposes of PP development to display which scores have these differences from being in stable, and which scores are set in lazer and have the intentional changes.

A quick idea I came up with is to replace the mod "Classic" with the mod "Legacy" for stable scores on the leaderboard. When the replay is run, the score would be displayed as being a "Classic" score, as the play is now being simulated as if it were set in lazer, so there wouldn't need to be any new gameplay behaviour implemented to account for this change. It doesn't even have to be a mod, a tooltip or an icon somewhere on the score panel would suffice.

![image](https://user-images.githubusercontent.com/92956514/221387602-c4c6260b-2789-4d61-85e3-6d993ab33661.png)

**Comment by bdach** [comment]
Can we not fragment things further than they already are going to be already? What is the point of the entire "classic" exercise if you're going to pile "legacy" onto the mix?

Can we start with describing the actual problem you want to solve here? It seems to be this:

> Classic scores set in lazer have differences to scores set in stable, such as information on slider ends dropped and sliderticks missed, as well as not losing combo from a dropped sliderend

so maybe there are ways to backfill this data, at least for the important scores, rather than multiply entities and cause confusion?

> A quick idea I came up with is to replace the mod "Classic" with the mod "Legacy" for stable scores on the leaderboard. When the replay is run, the score would be displayed as being a "Classic" score, as the play is now being simulated as if it were set in lazer

This just makes my head hurt. How does this make things any easier to explain UX-wise if a mod just presto-change-o gets swapped for another one?

**Reply by Axilotl17** [reply]
classic is a mod to emulate playing on stable, but the idea for legacy is a tag for plays ACTUALLY set on stable. so leaderboards could have legacy scores (set on stable) and classic scores (set on lazer w/ classic mod)

**Reply by Natelytle** [reply]
> Can we start with describing the actual problem you want to solve here?

The idea was something I've been thinking of for almost the entirety of my time spent developing [#22613](https://github.com/ppy/osu/pull/22613), as there's no way to discern a legacy score from a classic mod score in the PP system, and it's a huge blocker for the PR getting merged as I've had to settle for a hacky workaround to account for it. The biggest issue with the mania situation is there has been no work done on a classic mod for mania for around 2 years now, as it's a massive undertaking and would definitely take a huge amount of effort to create one. As well, there are talks of per-note hit judgements, and many notable scores don't have any replays to backfill information from.

> How does this make things any easier to explain UX-wise if a mod just presto-change-o gets swapped for another one?

The mod is a very rough first pass idea on what could be done, and like I said in the original issue, it doesn't have to be a mod. A tooltip where the sliderend and slidertick information is would be perfectly fine, and there are definitely more UX friendly ways to achieve the same goal.

**Reply by plaguemask** [reply]
Just my two cents, but I think this discussion's existence shows that the current implementation of Classic mod is fundamentally flawed. It is portrayed as equivalent to a stable score, yet it is calculated from different data produced by different mechanics. Why label stable scores with Classic mod if the mod can't faithfully recreate such scores, as would be expected? But to make the mod do so would be contrary to the concept of lazer progressing past stable with its new scoring and gameplay mechanics.

I feel with how Classic mod is right now, lazer is trying to have its cake and eat it too. It wants to make bold changes, but also be a seamless transition. It wants to shed all of stable's quirks, but also fully incorporate its decade-plus-long history. Trying to do both of these at once is causing confusion that, while still minor for now, I feel will only worsen until one path is chosen, committed to, and communicated to the user. Either Classic mod needs to truly match stable's gameplay and scoring, or it needs to be made clear that will never be the case by not conflating stable-made scores and lazer-made scores.

**Reply by mk56-spn** [reply]
> Just my two cents, but I think this discussion's existence shows that the current implementation of Classic mod is fundamentally flawed. It is portrayed as equivalent to a stable score, yet it is calculated from different data produced by different mechanics. Why label stable scores with Classic mod if the mod can't faithfully recreate such scores, as would be expected? But to make the mod do so would be contrary to the concept of lazer progressing past stable with its new scoring and gameplay mechanics.
> 
> I feel with how Classic mod is right now, lazer is trying to have its cake and eat it too. It wants to make bold changes, but also be a seamless transition. It wants to shed all of stable's quirks, but also fully incorporate its decade-plus-long history. Trying to do both of these at once is causing confusion that, while still minor for now, I feel will only worsen until one path is chosen, committed to, and communicated to the user. Either Classic mod needs to truly match stable's gameplay and scoring, or it needs to be made clear that will never be the case by not conflating stable-made scores and lazer-made scores.

It's not that deep, the mod tries to emulate the feel of stable MUCH more closely than normal Lazer does. So it's called classic mode. It's hardly the first game that has " legacy" gameplay features in newer version / titles that don't exactly match up down to the last millimeter, often that's just not feasible. It's name describes that it does well enough

And as bdach said already the added confusion/ complexity from further separating the categories might not be worth it. It isn't in my opinion anyways but hey.

**Reply by peppy** [reply]
> classic is a mod to emulate playing on stable, but the idea for legacy is a tag for plays ACTUALLY set on stable. so leaderboards could have legacy scores (set on stable) and classic scores (set on lazer w/ classic mod)

The whole point of classic mod is to make scores set with it comparable to stable. While we may add something in the UI showing whether a score was set to classic or stable, it will not be a separate mod as proposed in this thread. If we can't get classic mod to the point that scores are roughly comparable then it should probably be removed along with all display of legacy scores in the lazer client forever (which is hopefully not the direction we need to go).

If you're interested in more discussion on this matter, please start by watching or reading through the last community meeting notes.

**Reply by Natelytle** [reply]
Keep in mind that no matter how roughly comparable to stable scores set in lazer with the classic mod are, there will always be situations where the PP system needs to know whether a score was set in stable vs lazer to properly weight it. For example, since lazer classic mod doesn't make you lose combo for a dropped sliderend, the sliderend estimation code in the PP system will have to know whether or not it is necessary, and classic mod doesn't provide this information. Also reiterating on my point about mania, there has been 0 work on a mania classic mod for ~3 years, and it's definitely not looking like a high priority, however knowing which scores were set with classic behaviour is a must in order to balance scores set on stable vs lazer.

**Reply by peppy** [reply]
I can only reiterate: work will sooner be done to match the two. Not to differentiate the two.
